### PR TITLE
Secrets store hot backup and restore

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
Allow users to configure a "backup" secrets store in addition to the primary one, with the same set of available secrets store types and configuration options.

When the backup secrets store is configured, it acts as a backup and high-available replica of the primary secrets store:
* all secret values are also stored (created/updated) in the backup secrets store
* if the primary secrets store is not responding or failing, the backup is used to fetch, create and delete secret values (the update operation is the only one impacted, for consistency reason).
* if the backup secrets store is not responding or failing, the errors are treated as warnings
* if a secret value is *only* found in the backup secrets store, it is automatically copied back to the primary secrets store on retrieval
* two new `zenml secret backup` and `zenml secret restore` CLI commands allow bulk secret backup and restoration (see below)

This feature is also used to migrate secrets between secrets stores in three phases:
1. first, the new secrets store (B) is configured as backup alongside the existing one (A)
2. a bulk secret store values backup operation is called by running `zenml secret backup --delete-secrets --ignore-errors`. This effectively moves all secret values from secrets store (A) to (B).
3. the primary secret store (A) is removed from the configuration and the new one (B) is configured as primary.

### Side-quests

* new REST API endpoints are implemented for the new backup/restore bulk secret operations
* the new backup/restore bulk secret operations make use of two new RBAC secret permissions that need to be reflected in the ZenML Cloud code & dashboard. Secret BACKUP and RESTORE permissions should only be assigned to the tenant admin by default.
* this PR also re-organizes the helm chart templates to generate environment variables using named templates instead of duplicating them in various places. This eliminates code duplication and keeps the helm chart code closer aligned to the ZenML python configuration code.
* the GCP service account credentials, when configured for the GCP secrets store, are no longer passed as a pod volume, but as an environment variable instead.
* improved documentation for secrets stores and docker/helm deployment configuration options related to secrets stores

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

## Summary by CodeRabbit

- **New Features**
  - Introduced backup and restore commands for secrets management.
  - Added support for configuring a backup secrets store via environment variables.
  - New API endpoints for secret operations including backup and restore functionalities.

- **Enhancements**
  - Enhanced secret stores to handle exceptions more explicitly for better error reporting.
  - Improved clarity of error messages in secret retrieval across various secret stores.

- **Documentation**
  - Updated RBAC models to include new `BACKUP` and `RESTORE` actions.

### CODE REVIEW

**Comment by stefannica** [comment]
@coderabbitai review

File: src/zenml/zen_stores/zen_store_interface.py [hunk 1]
```diff
@@ -1456,6 +1456,42 @@
...
+        Args:
+            ignore_errors: Whether to ignore individual errors during the
+                restore process and attempt to restore all secrets.
+            delete_secrets: Whether to delete the secrets that have been
+                successfully restored from the backup secrets store. Setting
+                this flag effectively moves all secrets from the backup secrets
+                store to the primary secrets store.
+
+        Raises:
+            NotImplementedError: if no backup secrets store is configured.
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_stores/zen_store_interface.py]: done

File: src/zenml/zen_stores/zen_store_interface.py [hunk 2]
```diff
@@ -1456,6 +1456,42 @@ def delete_secret(self, secret_id: UUID) -> None:
...
+            ignore_errors: Whether to ignore individual errors during the backup
+                process and attempt to backup all secrets.
+            delete_secrets: Whether to delete the secrets that have been
+                successfully backed up from the primary secrets store. Setting
+                this flag effectively moves all secrets from the primary secrets
+                store to the backup secrets store.
+
+        Raises:
+            NotImplementedError: if no backup secrets store is configured.
+        """
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_stores/zen_store_interface.py]: done

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/zen_server/rbac/models.py [hunk 3]
```diff
@@ -37,6 +37,12 @@ class Action(StrEnum):
     # Models
     PROMOTE = "promote"
 
+    # Secrets
+    BACKUP = "backup"
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_server/rbac/models.py]: As discussed, I merged them into one permission, `BACKUP_RESTORE`.

**Review by schustmi — APPROVED** [review]

File: src/zenml/zen_stores/sql_zen_store.py [hunk 4]
```diff
@@ -3851,33 +3887,163 @@ def _check_sql_secret_scope(
...
+                logger.exception(
+                    f"Failed to get secret values for secret with ID "
+                    f"{secret_id} from the primary secrets store. "
+                    f"Trying to get them from the backup secrets store. "
+                )
+                try:
+                    backup_values = self._get_backup_secret_values(
+                        secret_id=secret_id
+                    )
+                    if isinstance(e, KeyError):
```
Comment (wjayesh) [inline src/zenml/zen_stores/sql_zen_store.py:3996]: Beautiful!

File: src/zenml/zen_server/deploy/helm/templates/server-db-job.yaml [hunk 5]
```diff
@@ -61,68 +52,21 @@ spec:
...
-            - name: ZENML_SECRETS_STORE_VAULT_ADDR
-              value: {{ .Values.zenml.secretsStore.hashicorp.vault_addr | quote }}
-            {{- if .Values.zenml.secretsStore.hashicorp.vault_namespace }}
-            - name: ZENML_SECRETS_STORE_VAULT_NAMESPACE
-              value: {{ .Values.zenml.secretsStore.hashicorp.vault_namespace | quote }}
-            {{- end }}
-            {{- if .Values.zenml.secretsStore.hashicorp.max_versions }}
-            - name: ZENML_SECRETS_STORE_MAX_VERSIONS
-              value: {{ .Values.zenml.secretsStore.hashicorp.max_versions | quote }}
+            {{- range $k, $v := include "zenml.serverEnvVariables" . | fromYaml }}
```
Comment (wjayesh) [inline src/zenml/zen_server/deploy/helm/templates/server-db-job.yaml:59]: The one concern I had when I first saw that you replaced the google credentials mount with env vars is that the libraries won't work without having a file reference with the credentials. I quickly realised that we have now moved to using service connectors instead; neat!

**Review by wjayesh — APPROVED** [review]
Looking forward to having this merged, great job!
Also, for my understanding, the way to switch a secrets store from backup to primary is just by redeploying the server with the backup store configuration as the primary one, correct?

**Pull request merged**